### DISCUSSION TITLE
Compiler.watch bug

### DISCUSSION THREAD

**Discussion by ygs-code** [webpack/webpack#15901] [Q&A]
# Bug report

**What is the current behavior?**

**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**

**Other relevant information:**
webpack version:  
Node.js version: 
Operating System: 
Additional tools:

Hello, I'm reporting an error here  ，You ran Webpack twice. Each instance only supports a single concurrent compilation at a time.
    at Compiler.watch 。 I have a plugin that NEEDS to be called Compiler.watch  Get the compilation information。
My plug-in code looks like this。 If the solution，You ran Webpack twice. Each instance only supports a single concurrent compilation at a time.
    at Compiler.watch？

const name = "BrowserReloadErrorOverlayWepbackPlugin";

class BrowserReloadErrorOverlayWepbackPlugin {
  constructor(options) {
    this.options = {
      port: 8080,
      retryWait: 5000,
      delay: 0,
      ...options,
    };
  }

  async start() {
    const { options } = this;
    let port = options.port;
    // 设置静态服务器
    portfinder.basePort = port;
    port = await new Promise((resolve, reject) => {
      //查找端口号
      portfinder.getPort((err, port) => {
        if (err) {
          reject(err);
          return;
        }
        // 新端口
        resolve(port);
      });
    });

    this.wss = new WebSocket.Server({ port });

    this.wss.on("connection", function (ws) {
      // 远程socket
      // this.broadcast(
      //   JSON.stringify({
      //     cmd: "cmd:reload",
      //     message: "",
      //   })
      // );
    });

    this.clientSrc = createClient({
      port,
      retryWait: options.retryWait,
      delay: options.delay,
    });
  }

  broadcast(message) {
    this.wss.clients.forEach((client) => client.send(message));
  }
  // // 做兼容
  hook(compiler, hookName, pluginName, fn) {
    if (arguments.length === 3) {
      fn = pluginName;
      pluginName = hookName;
    }
    if (compiler.hooks) {
      compiler.hooks[hookName].tap(pluginName, fn);
    } else {
      compiler.plugin(pluginName, fn);
    }
  }
  apply(compiler) {
    const { options, clientSrc } = this;
    const { compilerWatch = () => {} } = options;
    // // // 开始编译 只会调用一次
    this.hook(compiler, "afterPlugins", async (compilation) => {
      if (!this.wss) {
        await this.start();
      }
    });

    // const matchObject = ModuleFilenameHelpers.matchObject.bind(undefined, {
    //   include: options.include || /\.js$/,
    //   exclude: options.exclude,
    // });

 
    compiler.watch(
      {
        // [watchOptions](/configuration/watch/#watchoptions) 示例
        // aggregateTimeout: 300,
        poll: undefined,
      },
      async (err, stats) => {
        compilerWatch(err, stats);
        
        this.broadcast(
          JSON.stringify({
            cmd: "cmd:rebuilding",
            message: "",
          })
        );
        // stabilization(500).then(() => {
        if (err) {
          this.broadcast(
            JSON.stringify({
              cmd: "cmd:buildError",
              message: err.stack || err || err.details,
            })
          );
          // console.log('Errors:' + chalk.red(err.stack || err));
          // if (err.details) {
          //     console.log('Errors:' + chalk.red(err.details));
          // }
          return;
        }

        if (stats.hasErrors()) {
          this.broadcast(
            JSON.stringify({
              cmd: "cmd:buildError",
              message:
                stats.toString({
                  colors: true,
                }) + "\n\n",
            })
          );
          // console.log(
          //     'Errors:' +
          //         chalk.red(
          //             stats.toString({
          //                 colors: true,
          //             }) + '\n\n'
          //         )
          // );
          return;
        } else if (stats.hasWarnings()) {
          // console.log(
          //   "Warnings:" +
          //     chalk.yellow(
          //       stats.toString({
          //         colors: true,
          //       }) + "\n\n"
          //     )
          // );
          // return;
        }

        this.broadcast(
          JSON.stringify({
            cmd: "cmd:reload",
            message: "",
          })
        );
        // else {
        //     process.stdout.write(
        //         stats.toString({
        //             colors: true,
        //         }) + '\n\n'
        //     );
        // }
      }
    );

    compiler.hooks.emit.tap("ErrorOverlayWebpack", (compilation) => {
      const { clientSrc } = this;
      for (const name in compilation.assets) {
        if (compilation.assets.hasOwnProperty(name) && name.endsWith(".js")) {
          const contents = compilation.assets[name].source();
          const withoutComments =
            ansiToHtml + "\n" + clientSrc + "\n" + contents;
          compilation.assets[name] = {
            source: () => withoutComments,
            size: () => withoutComments.length,
          };
        }
      }
    });
  }
}

**Comment by vankop** [comment]
I don't understand what you trying to achieve... please fill issue template.

**Comment by ygs-code** [comment]
My problem is the same as his https://stackoverflow.com/questions/65278463/watching-mode-in-webpack-node-api/65279119

**Comment by ygs-code** [comment]
let compiler = webpack(webpackConfig);
compiler.watch({}, (err, stats) => {
  if (err || stats.hasErrors()) console.error(err);
  console.log('Webpack refresh...');
});             I configured it so that launching webPack would generate an error message，ConcurrentCompilationError: You ran Webpack twice. Each instance only supports a single concurrent compilation at a time.。Watch is fine if I don't use compiler.watch, but I need to use the compiler.watch method。You know what I'm saying？

**Comment by vankop** [comment]
you should not provide watch options in `webpack.config` then, use them in `compiler.watch` as an argument

**Reply by ygs-code** [reply]
My configuration is going to look like this。

const BrowserReloadErrorOverlayWepbackPlugin = require('browser-reload-error-overlay-wepback-plugin');

let compiler = webpack({

// ... Omit the configuration
plugins: [
new BrowserReloadErrorOverlayWepbackPlugin(),
]

});

BrowserReloadErrorOverlayWepbackPlugin npm

https://www.npmjs.com/package/browser-reload-error-overlay-wepback-plugin

Main error code
I used it in the plugin compiler.watch

apply(compiler) {
   compiler.watch({},(err, stats) => {
   })
}

And then there was this mistake:
ConcurrentCompilationError: You ran Webpack twice. Each instance only supports a single concurrent compilation at a time.
at Compiler.watch (K:\webpack-cli\node_modules\webpack\lib\Compiler.js:411:19)
at webpack (K:\webpack-cli\node_modules\webpack\lib\webpack.js:144:15)
at f (K:\webpack-cli\node_modules\webpack\lib\index.js:63:16)
at App.setCompiler (K:\webpack-cli@webpack-cli\compiler-server.js:96:26)
at App.middleware (K:\webpack-cli@webpack-cli\compiler-server.js:445:29)
at App.init (K:\webpack-cli@webpack-cli\compiler-server.js:67:14)
at processTicksAndRejections (node:internal/process/task_queues:96:5)

**Reply by vankop** [reply]
`compiler.watch` should not be called in plugin
